### ISSUE TITLE
dump

### ISSUE THREAD

**Issue by valiantlynx** [kuben-labs/nmai#1]
Files Created (7 files)
File	Purpose
AGENTS.md	~150-line guide for AI agents with all commands, style rules, constraints
configs/dataset.yaml	YOLOv8 dataset config with all 356 category names
scripts/coco_to_yolo.py	COCO-to-YOLO conversion (already run, data/yolo/ populated)
scripts/train.py	Training launcher with tuned hyperparameters for this dataset
scripts/run.py	Submission inference script (sandbox entry point)
scripts/evaluate.py	Local competition score calculator (det + cls mAP)
scripts/validate_predictions.py	Predictions format validator
scripts/package_submission.py	Zip builder with all limit checks

Data Prepared
- data/yolo/ -- YOLO-format dataset with 211 train / 37 val images (symlinked) and labels
- All 22,731 annotations converted and verified (spot-checked: exact match with COCO GT)

How to Run the Full Pipeline
# 1. Install (on your GPU machine)
pip install ultralytics==8.1.0
pip install torch==2.6.0 torchvision==0.21.0 --index-url https://download.pytorch.org/whl/cu124
pip install pycocotools==2.0.7 albumentations==1.3.1
# 2. Train (start here -- ~30 min baseline)
python scripts/train.py --model yolov8m.pt --imgsz 640 --epochs 50 --name baseline
# 3. Evaluate locally
python scripts/evaluate.py --model runs/baseline/weights/best.pt \
    --images data/yolo/images/val --val-only
# 4. Scale up once baseline works
python scripts/train.py --model yolov8m.pt --imgsz 1280 --epochs 150 --name full
# 5. Package and submit
python scripts/package_submission.py --weights runs/full/weights/best.pt

Training Script Highlights
The train.py hyperparameters are tuned for this specific dataset:
- AdamW + cosine LR for stable training on small datasets
- scale=0.5 to handle the wild image size variation (481-5712px)
- mosaic + mixup + copy_paste for aggressive augmentation on only 211 training images
- close_mosaic=15 disables mosaic for last 15 epochs (refines predictions)
- patience=30 for early stopping
- AMP enabled for mixed-precision training speed

**Comment by valiantlynx** [comment]
Nothing special was done here. It's a straightforward YOLOv8 fine-tuning pipeline. Let me walk through the whole thing.
What was built
The problem
A competition to detect and classify grocery products on store shelf images. You get scored on:
- 70%: Did you find the products? (bounding box detection)
- 30%: Did you identify which product? (356 categories)
The approach
Fine-tune a pre-trained YOLOv8m model on the 248 labeled training images. YOLOv8 does detection and classification jointly in a single pass -- no separate classifier needed.
What each file does
scripts/coco_to_yolo.py -- The training annotations are in COCO format (JSON with [x, y, width, height] bounding boxes). YOLOv8 expects a different format (text files with normalized [center_x, center_y, width, height]). This script converts between the two and splits 248 images into 211 train / 37 validation. It uses symlinks instead of copying images to save disk space.
configs/dataset.yaml -- Tells YOLOv8 where the images/labels are and lists all 356 category names. Auto-generated by the conversion script with absolute paths for whatever machine it runs on.
scripts/train.py -- Launches YOLOv8 training with hyperparameters tuned for this dataset:
- AdamW + cosine LR: standard modern optimizer setup
- Heavy augmentation (mosaic, mixup, copy_paste, scale=0.5): because 211 training images is very small, so we need to create visual diversity artificially
- close_mosaic=15: turns off mosaic augmentation for the last 15 epochs so the model refines on clean images
- patience=30: stops training if validation mAP doesn't improve for 30 epochs
- torch.load monkey-patch: PyTorch 2.6 changed a default that breaks ultralytics 8.1.0's model loading. The patch restores the old behavior.
scripts/run.py -- The submission inference script. Loads the ONNX model, runs it on each test image, converts the output boxes back to COCO format, writes JSON. Uses ONNX instead of PyTorch because the sandbox's PyTorch has the weights_only incompatibility with ultralytics.
scripts/export_onnx.py -- Converts the trained .pt weights to ONNX format. ONNX is a framework-neutral model format that onnxruntime can load without any torch.load issues.
scripts/evaluate.py -- Computes the competition score locally so you can iterate without submitting. Reimplemented mAP computation from scratch because pycocotools had a numpy binary incompatibility on the training machine.
scripts/validate_predictions.py -- Checks that the output JSON has the right format before you waste a submission.
scripts/package_submission.py -- Copies run.py + weights into a zip with the right structure (run.py at root, not in a subfolder).
The results
Run	Model	Image size	Epochs	mAP@0.5 (mailto:mAP@0.5)
baseline	yolov8m	640	50	0.624
full2	yolov8m	1280	117 (early stop at 87)	0.799
The improvement from baseline to full came from two things: larger input images (640 -> 1280, which matters because products are small on shelves) and more training epochs.
What went wrong along the way
1. PyTorch 2.6 weights_only=True -- The training machine has PyTorch 2.6 which changed torch.load defaults. ultralytics 8.1.0 doesn't handle this. Fixed with a monkey-patch.
2. numpy 2.4.3 incompatibility -- pip kept upgrading numpy when reinstalling pycocotools, which broke both pycocotools (C extension compiled against different numpy) and ultralytics (np.trapz was removed in numpy 2.0). Fix: pin numpy==1.26.4.
3. Sandbox exit code 1 -- The .pt submission failed because the sandbox has the same PyTorch 2.6 weights_only issue, and monkey-patching torch.load in the submission script is risky with security scanners. Fix: export to ONNX, which sidesteps torch.load entirely.
4. runs/full vs runs/full2 -- ultralytics auto-increments the run name if the directory already exists. Minor annoyance, not a bug.
Nothing clever
There's no novel technique here. It's: take pre-trained model, convert data format, fine-tune, export, submit. The hyperparameters are standard recommendations from the ultralytics docs adjusted for a small dataset (more augmentation, smaller learning rate). The ONNX export is the documented way to handle version mismatches.
